[@rails/activestorage] Update types for @rails/activestorage v8.0

victoria-huang opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/rails/rails/blob/17cde11cd13cf07a7d6c712cb1593881496abacb/activestorage/app/javascript/activestorage/direct_upload.js#L32-L38
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

This updates the DirectUpload create callback error argument type to be nullable and the blob argument type to be optional to more closely match the Rails JavaScript implementation. The [`error` argument is `null` when the direct upload succeeds, and the `blob` argument is not present when there is an `error`](https://github.com/rails/rails/blob/17cde11cd13cf07a7d6c712cb1593881496abacb/activestorage/app/javascript/activestorage/direct_upload.js#L32-L38):

```js
upload.create(error => {
  if (error) {
    callback(error)
  } else {
    callback(null, blob.toJSON())
  }
})
```

imustafin commented:
Hi. Thanks for the PR!

Would you also make the second argument optional (`?`) for the case when `error` is present?

Should we update the typings version?

Tests would be nice to have.

Otherwise the PR is good.

victoria-huang commented:
Hi @imustafin, thanks for the feedback!

> Would you also make the second argument optional (`?`) for the case when `error` is present?

This makes sense to me - updated the `blob` argument to be optional.

> Should we update the typings version?

I updated the minor version in `package.json` to 7.2.

> Tests would be nice to have.

Added tests - please let me know if there's anything else I can do here.

imustafin reviewed (changes requested):
It took me some time to reread TD's rules and activestorage's code.

One small fix for the version and we can merge!

victoria-huang commented on types/rails__activestorage/package.json:
```diff
@@ -1,7 +1,7 @@
     "private": true,
     "name": "@types/rails__activestorage",
-    "version": "7.1.9999",
+    "version": "7.2.9999",
```
Thanks @imustafin. I was admittedly unsure if this should be a major version change, but using the newest version makes sense to me. I am currently using activestorage v7 but will be upgrading to v8 - in testing, I can confirm that these changes work for both on my end.

imustafin reviewed (approved):
Thanks!

victoria-huang commented:
Ready to merge
